[wip] use DimVector for sizes and strides

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53001 [wip] use DimVector for sizes and strides**

This PR implements @ngimel's suggestion to apply @swolchok's `unsqueeze` optimization in #49497 to `view`, i.e. use `SmallVector`s instead of `std::vector`s to store intermediate shape and stride values. `v = t.view(-1)` in a tight loop shows instruction count reduction of ~15%.

Reviewers: this PR is straightforward, but has some incidental complexity due to the multiple representations in play for sizes and strides. The simplest reading sequence is

1. [TensorShape.cpp](https://github.com/pytorch/pytorch/pull/53001/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cf) has the changed implementation of `view()`, along with changes to `alias_with_sizes_and_strides()`, which it calls. That function has been templatized to generate a version for both `IntArrayRef`s and `SmallVector<int64_t>`s, which are API compatible. This templating approach is repeated below for other helpers. 

2. `InferSize.h` and `TensorUtils.h/cpp` both extend functions called from `view()` to handle the `SmallVector` use case. Wrinkles in each of these cases prevent us from doing a straight templatization of the original function, instead we factor out a helper and templatize that. Pretty straightforward.

Note: in general, the `IntArrayRef` dataflows through these helpers are sourced from "persistent" sizes and strides hosted by Tensors, whereas the `SmallVector` data flows are sourced from ephemeral computed sizes and strides (prior to this diff, these were stored in `std::vector` and implicitly converted to `IntArrayRef`). So the need to handle dual representations in helpers isn't going away. 

Also though, there are certainly opportunities for improving the performance of other ops that generate intermediate size and strides data by using the `SmallVector`-enabled helpers introduced here (or extending other helpers similarly.)

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

Commit:
use DimVector for sizes and strides

[ghstack-poisoned]

Commit:
Update on "use DimVector for sizes and strides"


Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "use DimVector for sizes and strides"


Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


We can save quite a bit by using `SmallVector<long, 5>` aka `DimVector` for size and stride computations - the current version of this PR does this on the `t.view()` callpath and saves ~13% in instruction count.

Not ready for review - other clients of size and stride calculation code need to be ported.

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

swolchok commented on aten/src/ATen/InferSize.h:
```diff
@@ -41,12 +46,24 @@ inline std::vector<int64_t> infer_size(IntArrayRef shape, int64_t numel) {
       res[*infer_dim] = numel / newsize;
     }
-    return res;
+    return;
```
nit: no need to return at the end of a void function

bhosmer replied:
true, but this isn't at the end :)

swolchok commented on aten/src/ATen/InferSize.h:
```diff
@@ -41,12 +46,24 @@ inline std::vector<int64_t> infer_size(IntArrayRef shape, int64_t numel) {
 
+inline std::vector<int64_t> infer_size(IntArrayRef shape, int64_t numel) {
+  auto res = shape.vec();
+  infer_size_impl(res, shape, numel);
```
nit: putting the out parameter last means you won't have to shift arguments around if somehow infer_size and/or infer_size_impl don't get inlined

swolchok reviewed (approved):
lgtm

Commit:
Update on "[wip] use DimVector for sizes and strides"


This PR implements @ngimel's suggestion to apply @swolchok's `unsqueeze` optimization in #49497 to `view`, i.e. use `SmallVector`s instead of `std::vector`s to store intermediate shape and stride values. `v = t.view(-1)` in a tight loop shows instruction count reduction of ~15%.

Reviewers: this PR is straightforward, but has some incidental complexity due to the multiple representations in play for sizes and strides. The simplest reading sequence is

1. [TensorShape.cpp](https://github.com/pytorch/pytorch/pull/53001/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cf) has the changed implementation of `view()`, along with changes to `alias_with_sizes_and_strides()`, which it calls. That function has been templatized to generate a version for both `IntArrayRef`s and `SmallVector<int64_t>`s, which are API compatible. This templating approach is repeated below for other helpers. 

2. `InferSize.h` and `TensorUtils.h/cpp` both extend functions called from `view()` to handle the `SmallVector` use case. Wrinkles in each of these cases prevent us from doing a straight templatization of the original function, instead we factor out a helper and templatize that. Pretty straightforward.

Note: in general, the `IntArrayRef` dataflows through these helpers are sourced from "persistent" sizes and strides hosted by Tensors, whereas the `SmallVector` data flows are sourced from ephemeral computed sizes and strides (prior to this diff, these were stored in `std::vector` and implicitly converted to `IntArrayRef`). So the need to handle dual representations in helpers isn't going away. 

Also though, there are certainly opportunities for improving the performance of other ops that generate intermediate size and strides data by using the `SmallVector`-enabled helpers introduced here (or extending other helpers similarly.)

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

Commit:
Update on "[wip] use DimVector for sizes and strides"


This PR implements @ngimel's suggestion to apply @swolchok's `unsqueeze` optimization in #49497 to `view`, i.e. use `SmallVector`s instead of `std::vector`s to store intermediate shape and stride values. `v = t.view(-1)` in a tight loop shows instruction count reduction of ~15%.

Reviewers: this PR is straightforward, but has some incidental complexity due to the multiple representations in play for sizes and strides. The simplest reading sequence is

1. [TensorShape.cpp](https://github.com/pytorch/pytorch/pull/53001/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cf) has the changed implementation of `view()`, along with changes to `alias_with_sizes_and_strides()`, which it calls. That function has been templatized to generate a version for both `IntArrayRef`s and `SmallVector<int64_t>`s, which are API compatible. This templating approach is repeated below for other helpers. 

2. `InferSize.h` and `TensorUtils.h/cpp` both extend functions called from `view()` to handle the `SmallVector` use case. Wrinkles in each of these cases prevent us from doing a straight templatization of the original function, instead we factor out a helper and templatize that. Pretty straightforward.

Note: in general, the `IntArrayRef` dataflows through these helpers are sourced from "persistent" sizes and strides hosted by Tensors, whereas the `SmallVector` data flows are sourced from ephemeral computed sizes and strides (prior to this diff, these were stored in `std::vector` and implicitly converted to `IntArrayRef`). So the need to handle dual representations in helpers isn't going away. 

Also though, there are certainly opportunities for improving the performance of other ops that generate intermediate size and strides data by using the `SmallVector`-enabled helpers introduced here (or extending other helpers similarly.)

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2070,7 +2074,7 @@ Tensor numpy_T(const Tensor &self) {
 
 Tensor view(const Tensor& self, IntArrayRef size) {
-  auto inferred_size = at::infer_size(size, self.numel());
+  at::DimVector inferred_size = at::infer_size_dv(size, self.numel());
```
can you make `unflatten` and `reshape` use `infer_size_dv` and get rid of `infer_size`?

bhosmer replied:
Yes to the former (done!), sadly no to the latter - there are other callers. (The water is muddied by another `infer_size()` defined in `ExpandUtils.h/cpp`: some callers of our infer_size() wind up resolving incorrectly to that one if we remove ours, making it look like there are no callers to ours outside this file. But renaming our `DimVector`-returning `infer_size_dv()` to `infer_size()` provokes compile errors that reveal the callers - https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/static/ops.cpp#L29, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mkldnn/TensorShape.cpp#L47. Those are entangled in a way that will make them hard to port over.)

Commit:
Update on "[wip] use DimVector for sizes and strides"


This PR implements @ngimel's suggestion to apply @swolchok's `unsqueeze` optimization in #49497 to `view`, i.e. use `SmallVector`s instead of `std::vector`s to store intermediate shape and stride values. `v = t.view(-1)` in a tight loop shows instruction count reduction of ~15%.

Reviewers: this PR is straightforward, but has some incidental complexity due to the multiple representations in play for sizes and strides. The simplest reading sequence is

1. [TensorShape.cpp](https://github.com/pytorch/pytorch/pull/53001/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cf) has the changed implementation of `view()`, along with changes to `alias_with_sizes_and_strides()`, which it calls. That function has been templatized to generate a version for both `IntArrayRef`s and `SmallVector<int64_t>`s, which are API compatible. This templating approach is repeated below for other helpers. 

2. `InferSize.h` and `TensorUtils.h/cpp` both extend functions called from `view()` to handle the `SmallVector` use case. Wrinkles in each of these cases prevent us from doing a straight templatization of the original function, instead we factor out a helper and templatize that. Pretty straightforward.

Note: in general, the `IntArrayRef` dataflows through these helpers are sourced from "persistent" sizes and strides hosted by Tensors, whereas the `SmallVector` data flows are sourced from ephemeral computed sizes and strides (prior to this diff, these were stored in `std::vector` and implicitly converted to `IntArrayRef`). So the need to handle dual representations in helpers isn't going away. 

Also though, there are certainly opportunities for improving the performance of other ops that generate intermediate size and strides data by using the `SmallVector`-enabled helpers introduced here (or extending other helpers similarly.)

Differential Revision: [D26719508](https://our.internmc.facebook.com/intern/diff/D26719508)

[ghstack-poisoned]
